[bandit] Validate args before subprocess calls

EddyMM opened the pull request:
### Description of changes
*  Validate device name arguments against a regex pattern before using it in subprocess call
   * N/B: The `manageVolume` and `ec2_dev_2_volid ` scripts are created in different directories (`/usr/local/sbin` and `/sbin` respectively) hence the use of their own independent validations (until we find a way to share functions, constants, modules between Python file in the cookbook)
* Add comments on why security exclusions remain valid (e.g. some subprocess calls are safe since they do not use any external arguments)


### Tests
* Unit tests
* Running existing Integrations tests

### References
* [[3.x] Reducing risks associated to Bandit findings #489](https://github.com/aws/aws-parallelcluster-node/pull/489)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

gmarciani commented on cookbooks/aws-parallelcluster-install/files/default/ec2_udev_rules/ec2_dev_2_volid.py:
```diff
@@ -24,11 +25,27 @@ def get_imdsv2_token():
+    """
+    match = re.match(r"^(\w){2,}$", device_name)
+    if not match:
+        raise ValueError("Device name provided argument contains has an invalid pattern.")
```
Typo: contains + has

gmarciani commented on cookbooks/aws-parallelcluster-install/files/default/ec2_udev_rules/ec2_dev_2_volid.py:
```diff
@@ -24,11 +25,27 @@ def get_imdsv2_token():
+    :raise: Exception if the argument fails to match the patter
+    :return: True if the argument matches the pattern
+    """
+    match = re.match(r"^(\w){2,}$", device_name)
```
Why` {2,}`?

EddyMM replied:
The block device names all have at least 2 alphanumeric chars [persistent_block_device_naming](https://wiki.archlinux.org/title/persistent_block_device_naming). For security purposes though, this regex constraint isn't really a hard requirement though. I can remove it.

gmarciani replied:
Since the goal of this change is security enforcement I would avoid introducing formatting constraints that may be out of scope and introduce risk (e.g. what if the name format changes in future? not highly probably I'd say but still can be avoided)

gmarciani commented on cookbooks/aws-parallelcluster-install/files/default/ec2_udev_rules/manageVolume.py:
```diff
@@ -15,6 +16,21 @@
 METADATA_REQUEST_TIMEOUT = 60
 
 
+def validate_device_name(device_name):
```
Why duplicating this logic (an almost equal logic has been defined in `cookbooks/aws-parallelcluster-install/files/default/ec2_udev_rules/ec2_dev_2_volid.py`.

What about putting the logic in common with parameter about the prefix of the accepted pattern?

EddyMM replied:
I agree. At the moment, these scripts are not in the form of a library (or in the same folder at execution time).

From the PR description:
The `manageVolume` and `ec2_dev_2_volid ` scripts are created in different directories (`/usr/local/sbin` and `/sbin` respectively) hence the use of their own independent validations (until we find a standard way to share functions, constants, modules between Python file in the cookbook)

EddyMM replied:
Some of the Python scripts in the cookbook are built as standalone scripts (with little/no code sharing capabilities).
Some suggestions to address this would be placing them in the same directory (which fixes the problem for scripts that can be in the same folder), or having a shared module installed in the virtualenv.

Either way, adding these changes seemed a bit too invasive for a simple fix to address a bandit finding. A cookbook-wide solution will need to be evaluated.

gmarciani replied:
Ok, considering the urgency of this Pr and the refactoring that this comment would imply, let's consider this as out of scope for this PR.

gmarciani commented on cookbooks/aws-parallelcluster-install/files/default/ec2_udev_rules/ec2_dev_2_volid.py:
```diff
@@ -24,11 +25,27 @@ def get_imdsv2_token():
     return headers
 
 
+def validate_device_name(device_name):
```
Is the name of this file correct? `ec2_dev_2_volid` shouldn't be `valid` instead?

EddyMM replied:
The name is correct `volid` -> Volume ID.
The script returns the volumen ID given the Block Device Name.

gmarciani replied:
ahahah ok thanks

EddyMM committed:
[bandit] Validate args before subprocess calls

- Validate device name arguments against a regex pattern before using it in subprocess call

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

EddyMM committed:
Merge branch 'develop' into wip/bandit-subprocess
